### DISCUSSION TITLE
How can we use efficiently Bertopic in another language?

### DISCUSSION THREAD

**Discussion by laurent-pincemaille** [MaartenGr/BERTopic#985] [General]
I tried Bertopic on 20newsgroup  dataset and was quite impressed by the results (since i was able to compair with the clustering made my Bertopic, i found really consistent topic quite similar to the original categories on my try).

I decided to reproduce it on  a french news corpus (similar to 20newsgroup dataset but in french) with the sentence-transformer  paraphrase-multilingual-MiniLM-L12-v2 as advised in https://maartengr.github.io/BERTopic/faq.html#why-are-the-results-not-consistent-between-runs. This time i got really poor result, most of my observations where in an anomaly cluster [and the algorithm seems to see a lot of cluster whereas i trained it on a 2 category  corpus and was expecting few clusters (sport and religion articles)]. 

Since there is specific model for English Embedding but for other languages we use undistinctly the same multilingual sentence-transformers i can understand that the result cannot bring the same quality. 

Is there a way to get a French embedding similar to the English ones (trained only on french corpus) ? 
On the repository of sentence-transformers: https://public.ukp.informatik.tu-darmstadt.de/reimers/sentence-transformers/v0.2/ it seems that there is only multilingual sentence-transformers, not specific ones ... Perhaps could you suggest me something similar in french or some tips with the multilingual model to perform well ?

PS: I tried to find something on internet and found only this one: https://huggingface.co/dangvantuan/sentence-camembert-large (the  topic_model.fit_transform(docs) fails when document are too long apparently)

**Comment by MaartenGr** [comment]
> This time i got really poor result, most of my observations where in an anomaly cluster [and the algorithm seems to see a lot of cluster whereas i trained it on a 2 category corpus and was expecting few clusters (sport and religion articles)].

There are a few things that might be happening here. First, HDBSCAN tends to find many clusters which you could prevent by either merging topics or by using k-Means and specifying k yourself since you mention that it should only contain 2 categories. In my experience, k-Means will handle fewer clusters a bit better than HDBSCAN. Second, the two categories in your dataset seem to be quite broad and as such, HDBSCAN will find sub-clusters and focus on local differences. Third, if you already have the categories, you could perform [supervised BERTopic](https://maartengr.github.io/BERTopic/getting_started/supervised/supervised.html) instead. Lastly, when using multi-lingual models, it might be worthwhile to additionally remove stopwords. 

> PS: I tried to find something on internet and found only this one: https://huggingface.co/dangvantuan/sentence-camembert-large (the topic_model.fit_transform(docs) fails when document are too long apparently)

It might be worthwhile to truncate your documents or split them up into sentences. That should prevent the errors that you are getting.

**Comment by drob-xx** [comment]
@laurent-pincemaille If you are trying to control the number of clusters created by HDBSCAN you might want to take a look at [TopicTuner](https://github.com/drob-xx/TopicTuner). It is fully compatible with BERTopic and will allow you to efficiently model your data to determine the best HDBSCAN parameters.

I thought I would also mention that in some cases I've found that a given corpus simply will not cluster well. In one case I had a 60K news story corpus that was basically breaking into 5-8 clusters which simply didn't make a lot of sense for what I was doing. On further inspection I found that the issue was that it was essentially breaking all the news into one set of clusters and all the sports into another. I did a clustering that identified all the "news" in one cluster and then separated that cluster from the others. Both the resultant news and sports corpi gave me very good results. They just didn't play well with one another. I played around with UMAP parameters to see if that would fix things - but in my case it didn't and splitting the corpus was a reasonable alternative.

**Comment by laurent-pincemaille** [comment]
Thanks a lot for all your advices (and sorry for the delayed answer). 
I finally tried https://huggingface.co/dangvantuan/sentence-camembert-large which performed quite well in fact (far better than Multilingual models) combined with  method "reduce_outliers" to decrease the number of outliers:
![image](https://user-images.githubusercontent.com/47352384/223384240-87ab2e58-5eec-42ea-b3d1-d27f3914603c.png)

At left you can see the ground Truth (MLSUM Dataset with categories as 20newsgroup dataset) and at right what Bertopic is performing. 
@MaartenGr you were right, we can see that Bertopic is even more precise as he can,for example, capture specific sports (football, ski, etc ... related articles) in a sport cluster. 

But as you can see there is a big blob (0 cluster) that i would like to split in distinct clusters. Is there a quick and easy way to deal with it ? To split a specific cluster ?
I was thinking to train a Bertopic model on this specific blob as you mentionned @drob-xx (so i first filter the documents on this specific blob and train a completely new Bertopic on it):
```
mask = df['topic_predicted']==0
df_bad_class =df[mask]
bad_docs = df_bad_class.text.to_list()
bad_embeddings = embeddings[mask]

topic_model2 = BERTopic(
  umap_model=umap_model,              # Step 2 - Reduce dimensionality
  hdbscan_model=hdbscan_model,        # Step 3 - Cluster reduced embeddings
  vectorizer_model=vectorizer_model,  # Step 4 - Tokenize topics
  ctfidf_model=ctfidf_model,          # Step 5 - Extract topic words
verbose=True,
  diversity=0.5                       # Step 6 - Diversify topic words
)
topics, probs = topic_model2.fit_transform(bad_docs, bad_embeddings)
```
but it is not really practical if i want to iterate this way a number of times (and the new topics number do not take in account the previous one). Is there an easier one to split and train a particular part of observations ?

**Reply by MaartenGr** [reply]
You can try to run a new model on that specific cluster to see if you can extract sub-clusters and pass those to a semi-supervised or manual instance of BERTopic. That way, you can find more fine-grained clusters whilst steering towards splitting up that larger cluster.

**Reply by laurent-pincemaille** [reply]
Thanks for your quick answer :).
Good idea! But i was wondering if doing this way won't affect the other possibilities that your package offert: such as reduce_outliers, or loosing track of the first topic assignation. Perhaps i should look at your partial_fit implementation since i suppose that it keep track of the different changes at each batch training ...

**Reply by MaartenGr** [reply]
With semi-supervised BERTopic you could still use everything BERTopic has to offer. With manual BERTopic, you could use most options, including reduce_outliers with the exception of just a few specific strategies. The idea behind the variations of BERTopic is that almost all functionality is kept.

**Reply by laurent-pincemaille** [reply]
Thanks a lot for your quick answers and your super nice package!
